xwiki-rendering-transformations/xwiki-rendering-transformation-macro/src/test/java/org/xwiki/rendering/internal/macro/DefaultMacroCategoryManagerTest.java

```diff
@@ -19,122 +19,181 @@
 {
-    private MacroCategoryManager macroCategoryManager;
+    @InjectMockComponents
+    private DefaultMacroCategoryManager macroCategoryManager;
 
-    @Override
-    @Before
-    public void setUp() throws Exception
-    {
-        super.setUp();
-        this.macroCategoryManager = getComponentManager().getInstance(MacroCategoryManager.class);
-    }
+    @MockComponent
+    private MacroTransformationConfiguration configuration;
+
+    @MockComponent
+    private MacroManager macroManager;
+
+    @RegisterExtension
+    LogCaptureExtension logCapture = new LogCaptureExtension(LogLevel.WARN);
 
     @Test
-    public void testGetMacroCategories() throws Exception
+    void getMacroCategories() throws Exception
     {
-        // TODO: This test needs to be improved. Right now it's based on the Test Macro located in the transformation
-        // package and for 4 of them a "Test" category has been set...
-        DefaultMacroTransformationConfiguration configuration =
-            (DefaultMacroTransformationConfiguration) getComponentManager().getInstance(
-                MacroTransformationConfiguration.class);
-        configuration.setCategories(new MacroId("testcontentmacro"), Set.of("Content"));
-        configuration.setCategories(new MacroId("testsimplemacro"), Set.of("Simple"));
+        Properties properties = new Properties();
+        properties.put("testcontentmacro", "Content");
+        properties.put("testsimplemacro", "Simple");
+        when(this.configuration.getCategories()).thenReturn(properties);
+        when(this.macroManager.getMacroIds())
+            .thenReturn(Set.of(new MacroId("testcontentmacro"), new MacroId("testsimplemacro")));
 
         Set<String> macroCategories = this.macroCategoryManager.getMacroCategories();
 
         // Check for a default category.
-        Assert.assertTrue(macroCategories.contains("Test"));
+        assertEquals(Set.of("Content", "Simple"), macroCategories);
+    }
 
-        // Check for null category.
-        Assert.assertTrue(macroCategories.contains(null));
+    @Test
+    void getMacroCategoriesNoCategory() throws Exception
+    {
+        MacroId macroAId = new MacroId("macroA");
+        DefaultMacroDescriptor macroADescriptor = new DefaultMacroDescriptor(macroAId, "Macro A", "Macro A");
+        Macro macroA = mock(Macro.class);
+
+        when(this.configuration.getCategories()).thenReturn(new Properties());
+        when(this.macroManager.getMacroIds()).thenReturn(Set.of(macroAId));
+        when(this.macroManager.getMacro(macroAId)).thenReturn(macroA);
+        when(macroA.getDescriptor()).thenReturn(macroADescriptor);
 
-        // Check for overwritten categories.
-        Assert.assertTrue(macroCategories.contains("Content"));
-        Assert.assertTrue(macroCategories.contains("Simple"));
+        // Check whether our macros are registered under correct categories.
+        Set<Object> expected = new HashSet<>();
+        expected.add(null);
```

Comment:
Why do we expect a set with `null`? Wouldn't an empty set be better?

Reply:
I've kept the logic where a macro without define category was affected to the "null" category. Without this special category, macros without category wouldn't be returned by some higher level operations.